001-collection-backend-dispatching.md

```diff
@@ -0,0 +1,165 @@
+- This feature does not propose a gpu (or RAPIDS) on/off switch. In the future, the backend defaults can be modified to depend on other hardware-preference information stored in the `dask.config` options. However, this feature only calls for backend configuration options at the collection level.
+- This feature does not mean Dask will take responsibility for testing a `cudf`-backed version of `dask.dataframe`. It is still the responsibility of the backend library (i.e. `cudf` and `dask_cudf`) to ensure that the collection and backend libraries are compatible.
+
+
+## Detailed Description
+
+
+### Designating the Backend (`dask.config`)
+
+From the perspective of the typical Dask user, the only visible result of the proposed feature is the addition of new fields in `dask/dask/dask.yaml`/`dask-schema.yaml` (accessiblke from `dask.config`). For each of the targeted collections (Dask-Array and Dask-DataFrame), we propose the addition of "backend", and "backend-options" fields. By default, the "backend" field will be set to "numpy" and "pandas" for Dask-Array and Dask-DataFrame, respectively. However, as shown in the code snippet below, this field can be changed with the existing `dask.config` interface to specify an alternative backend library.
+
+
+```python
+import dask
+
+with dask.config.set({"dataframe.backend": "cudf"}):
+    # Produce a cudf-backed collection
+    ddf = dask.dataframe.read_parquet("./tmpdir")
+
+with dask.config.set({"array.backend": "cupy"}):
+    # Produce a cupy-backed collection
+    darr = dask.array.ones(10, chunks=(5,))
+```
+
+Since it is unlikely that an alternative backend will support all numpy- or pandas-based data-creation functions available in the collection API, we also propose a "backend-options" field containing "allow-fallback" and "warn-fallback" subfields. When the "allow-fallback" field is set to `True` (default), then the backend's designated fallback class will be used to perform IO, and the result will be moved from the fallback backend. The user should also have the option to enable or disable warnings when this fallback behavior occurs:
+
+
+```python
+options = {
+    # Allow user to specify if a fallback backend library
+    # should be used, and if a warning should be raised when
+    # this occurs:
+    "allow-fallback" : True,
+    "warn-fallback" : True,
+}
+
+dask.config.set({"dataframe.backend-options": options})
+```
+
+The specific configuration options proposed here are certainly not set in stone.  However, we do feel that the user should have complete control over fallback behavior.  For example, the user should be able to specify if falling back to "numpy"/"pandas" should result in a warning or error message, or if it should be ignored altogether.
```

Comment:
I think hardcoding it is fine for now especially if we have in mind a mechanism for how it would work to support this kind of enhancement in the future.

Reply:
I added a note about this to the document.

Reply:
Note that I removed automated fallback behavior in the interest of a simpler design. The backend (e.g. cudf) can still implement their own fallback options if desired.